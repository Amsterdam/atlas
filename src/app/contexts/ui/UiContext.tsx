import { FunctionComponent } from 'react'
import { isEmbeddedParam } from '../../pages/MapPage/query-params'
import createNamedContext from '../../utils/createNamedContext'
import useParam from '../../utils/useParam'
import useRequiredContext from '../../utils/useRequiredContext'

export interface UiContextProps {
  /**
   * Indicates whether the application is being embedded through an iframe.
   */
  isEmbedded: boolean
}

const UiContext = createNamedContext<UiContextProps | null>('Ui', null)

export default UiContext

export const UiProvider: FunctionComponent = ({ children }) => {
  const [isEmbedded] = useParam(isEmbeddedParam)

  return <UiContext.Provider value={{ isEmbedded }}>{children}</UiContext.Provider>
}

export const useIsEmbedded = () => useRequiredContext(UiContext).isEmbedded
