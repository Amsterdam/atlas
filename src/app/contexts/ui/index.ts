export { default, UiProvider, useIsEmbedded } from './UiContext'
