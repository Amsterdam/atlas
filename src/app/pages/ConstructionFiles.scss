@import "../../shared/styles/config";

.c-construction-files {
  background-color: $primary-light;
  padding-bottom: 100px;

  &__title {
    margin-bottom: 7px !important;
  }

  .c-ds-table {
    margin-bottom: 10px;

    &__row {
      &:last-child {
        .c-ds-table__cell {
          border-bottom: 1px solid $secondary-gray20;
        }
      }
    }

    &__cell {
      padding-left: 20px;

      &:first-child {
        width: 33%;
      }
    }
  }
}
