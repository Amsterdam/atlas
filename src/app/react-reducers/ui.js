/* eslint-disable no-case-declarations */
export const REDUCER_KEY = 'ui'
export const SET_BACKDROP = `${REDUCER_KEY}/SET_BACKDROP`

export const initialState = {
  backdropKeys: [],
}

export default (state = initialState, action) => {
  switch (action.type) {
    // Increase or decrease the number of times the backdrop is triggered.
    // We do this so other triggers of show / hide backdrops won't overrule others while the
    // backdrop is active
    case SET_BACKDROP:
      const { key, open } = action.payload
      return {
        ...state,
        backdropKeys: state.backdropKeys.includes(key)
          ? state.backdropKeys.filter(backdropTrigger =>
              !open ? backdropTrigger !== key : backdropTrigger,
            )
          : [...state.backdropKeys, ...(open ? [key] : [])],
      }
    default:
      return state
  }
}

export const actions = {
  setBackDrop: SET_BACKDROP,
}
