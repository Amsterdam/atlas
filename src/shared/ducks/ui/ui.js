import { createSelector } from 'reselect';
import { isMapCurrentPage } from '../../../reducers/current-page-reducer';
import { getStraatbeeld } from '../straatbeeld/straatbeeld';

export const REDUCER_KEY = 'ui';

export const HIDE_EMBED_PREVIEW = 'HIDE_EMBED_PREVIEW';
export const HIDE_PRINT = 'HIDE_PRINT';
export const SHOW_EMBED_PREVIEW = 'SHOW_EMBED_PREVIEW';
export const SHOW_PRINT = 'SHOW_PRINT';
export const TOGGLE_MAP_PANEL_HANDLE = 'TOGGLE_MAP_PANEL_HANDLE';

const initialState = {
  isMapPanelHandleVisible: true,
  isEmbedPreview: false,
  isEmbed: false,
  isPrintMode: false
};

export default function UiReducer(state = initialState, action) {
  switch (action.type) {
    case HIDE_EMBED_PREVIEW:
      return {
        ...state,
        isEmbedPreview: false,
        isEmbed: false
      };

    case HIDE_PRINT:
      return {
        ...state,
        isPrintMode: false
      };

    case SHOW_EMBED_PREVIEW:
      return {
        ...state,
        isEmbedPreview: true,
        isEmbed: true
      };

    case SHOW_PRINT:
      return {
        ...state,
        isPrintMode: true
      };

    case TOGGLE_MAP_PANEL_HANDLE:
      return {
        ...state,
        isMapPanelHandleVisible: !state.isMapPanelHandleVisible
      };

    default:
      return state;
  }
}

// Todo: wire these actions properly when ui reducer is obsolete
export const toggleMapFullscreen = () => ({ type: 'NOOP' });
export const setMapFullscreen = () => ({ type: 'NOOP' });
export const toggleMapPanelHandle = () => ({ type: 'NOOP' });

// Selectors
const getUIState = (state) => state.ui;
export const isEmbedded = createSelector(getUIState, (ui) => ui.isEmbed);
export const isEmbedPreview = createSelector(getUIState, (ui) => ui.isEmbedPreview);
export const isPrintMode = createSelector(getUIState, (ui) => ui.isPrintMode);
export const isInPrintorEmbedMode = createSelector(
  isEmbedded,
  isPrintMode,
  isEmbedPreview,
  (embedded, print, preview) =>
    (embedded || print || preview));

// Todo: move this to a dataSelection ducks file
const getDataSelection = (state) => state.dataSelection || {};
const getDataSelectionView = createSelector(
  getDataSelection,
  (dataSelection) => dataSelection.view
);

export const isPrintModeLandscape = createSelector(
  isPrintMode,
  getStraatbeeld,
  isMapCurrentPage,
  getDataSelectionView,
  (printMode, straatbeeldActive, mapPageActive, dataSelectionView) =>
    (printMode && (!!straatbeeldActive || mapPageActive || (dataSelectionView === 'LIST')))
);

