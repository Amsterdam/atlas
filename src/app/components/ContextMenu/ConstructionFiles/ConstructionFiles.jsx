import React from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'
import { bindActionCreators } from 'redux'
import { ChevronDown, Ellipsis, Print, Download } from '@datapunt/asc-assets'
import { ContextMenu, ContextMenuItem, Icon } from '@datapunt/asc-ui'
import socialItems from '../socialItems'
import { sharePage, showPrintMode } from '../../../../shared/ducks/ui/ui'
import useDownload from '../../../utils/useDownload'
import getState from '../../../../shared/services/redux/get-state'

const ConstructionFiles = ({ openSharePage, fileName, fileUrl, openPrintMode, onDownload }) => {
  const { accessToken } = getState().user

  const [loading, downloadFile] = useDownload()
  const handleDownload = (trackEvent, imageUrl) => async () => {
    await downloadFile(imageUrl, {
      method: 'post',
      headers: new Headers({
        Authorization: `Bearer ${accessToken}`,
      }),
    })
    onDownload(trackEvent)
  }
  return (
    <ContextMenu
      data-test="context-menu"
      tabindex={0}
      alt="Actiemenu"
      arrowIcon={<ChevronDown />}
      icon={
        <Icon padding={4} inline size={24}>
          <Ellipsis />
        </Icon>
      }
      position="bottom"
    >
      <ContextMenuItem
        role="button"
        onClick={openPrintMode}
        icon={
          <Icon padding={4} inline size={24}>
            <Print />
          </Icon>
        }
      >
        Printen
      </ContextMenuItem>
      <ContextMenuItem
        as="button"
        disabled={loading}
        download={`${fileName}_small`}
        onClick={handleDownload('klein', `${fileUrl}/full/800,/0/default.jpg`)}
        icon={
          <Icon inline size={24} padding={4}>
            <Download />
          </Icon>
        }
      >
        Download klein
      </ContextMenuItem>
      <ContextMenuItem
        as="button"
        disabled={loading}
        download={`${fileName}_large`}
        onClick={handleDownload('groot', `${fileUrl}/full/1600,/0/default.jpg`)}
        icon={
          <Icon inline size={24} padding={4}>
            <Download />
          </Icon>
        }
      >
        Download groot
      </ContextMenuItem>
      <ContextMenuItem
        as="button"
        disabled={loading}
        download={`${fileName}_original`}
        divider
        onClick={handleDownload('origineel', `${fileUrl}/full/full/0/default.jpg`)}
        icon={
          <Icon inline size={24} padding={4}>
            <Download />
          </Icon>
        }
      >
        Download origineel
      </ContextMenuItem>
      {socialItems(openSharePage)}
    </ContextMenu>
  )
}

ConstructionFiles.propTypes = {
  openSharePage: PropTypes.func.isRequired,
  openPrintMode: PropTypes.func.isRequired,
  onDownload: PropTypes.func.isRequired,
  fileName: PropTypes.string.isRequired,
  fileUrl: PropTypes.string.isRequired,
}

const mapDispatchToProps = (dispatch) =>
  bindActionCreators(
    {
      openSharePage: sharePage,
      openPrintMode: showPrintMode,
    },
    dispatch,
  )

export default connect(null, mapDispatchToProps)(ConstructionFiles)
