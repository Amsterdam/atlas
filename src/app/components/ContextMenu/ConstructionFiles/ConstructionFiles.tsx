import { ChevronDown, Download, Ellipsis, Print } from '@amsterdam/asc-assets'
import { ContextMenu, ContextMenuItem, Icon } from '@amsterdam/asc-ui'
import React from 'react'
import { useDispatch } from 'react-redux'
import { sharePage, showPrintMode } from '../../../../shared/ducks/ui/ui'
import socialItems from '../socialItems'

type ConstructionFilesProps = {
  handleDownload: (imageUrl: string, size: string) => void
  fileUrl: string
  isImage: boolean
  downloadLoading: boolean
}

const ConstructionFiles: React.FC<ConstructionFilesProps> = ({
  fileUrl,
  handleDownload,
  downloadLoading,
  isImage,
}) => {
  const dispatch = useDispatch()

  const openSharePage = () => dispatch(sharePage())
  const openPrintMode = () => dispatch(showPrintMode())

  return (
    <ContextMenu
      data-test="context-menu"
      tabIndex={0}
      title="Actiemenu"
      arrowIcon={<ChevronDown />}
      icon={
        <Icon padding={4} inline size={24}>
          <Ellipsis />
        </Icon>
      }
      position="bottom"
    >
      <ContextMenuItem
        role="button"
        onClick={openPrintMode}
        icon={
          <Icon padding={4} inline size={24}>
            <Print />
          </Icon>
        }
      >
        Printen
      </ContextMenuItem>
      {isImage && (
        <>
          <ContextMenuItem
            disabled={downloadLoading}
            onClick={() => handleDownload(`${fileUrl}/full/800,/0/default.jpg`, 'klein')}
            icon={
              <Icon inline size={24} padding={4}>
                <Download />
              </Icon>
            }
          >
            Download klein
          </ContextMenuItem>
          <ContextMenuItem
            disabled={downloadLoading}
            onClick={() => handleDownload(`${fileUrl}/full/1600,/0/default.jpg`, 'groot')}
            icon={
              <Icon inline size={24} padding={4}>
                <Download />
              </Icon>
            }
          >
            Download groot
          </ContextMenuItem>
        </>
      )}
      <ContextMenuItem
        disabled={downloadLoading}
        divider
        onClick={() =>
          handleDownload(
            isImage ? `${fileUrl}/full/full/0/default.jpg` : `${fileUrl}?source_file=true`, // If the file is not an image the source file should be downloadable
            'origineel',
          )
        }
        icon={
          <Icon inline size={24} padding={4}>
            <Download />
          </Icon>
        }
      >
        Download origineel
      </ContextMenuItem>
      {socialItems(openSharePage)}
    </ContextMenu>
  )
}

export default ConstructionFiles
