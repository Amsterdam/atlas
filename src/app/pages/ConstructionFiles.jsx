import React from 'react';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import PropTypes from 'prop-types';
import { GridContainer, GridItem, Typography } from '@datapunt/asc-ui';
import { setCurrentFile } from '../../shared/ducks/files/actions';
import { getFileName } from '../../shared/ducks/files/selectors';
import { getUser } from '../../shared/ducks/user/user';
import { SCOPES } from '../../shared/services/auth/auth';
import NotAuthorizedMessage from '../components/PanelMessages/NotAuthorizedMessage';
import SHARED_CONFIG from '../../shared/services/shared-config/shared-config';
import { getLocationPayload } from '../../store/redux-first-router/selectors';
import Gallery from '../components/Gallery/Gallery';
import LoadingIndicator from '../../shared/components/loading-indicator/LoadingIndicator';
import ErrorMessage from '../components/PanelMessages/ErrorMessage/ErrorMessage';
import { getByUrl } from '../../shared/services/api/api';
import './ConstructionFiles.scss';
import { ConstructionFiles as ContextMenu } from '../components/ContextMenu';
import useMatomo from '../utils/useMatomo';
import useDocumentTitle from '../utils/useDocumentTitle';

const ImageViewer = React.lazy(() => import('../components/ImageViewer/ImageViewer'));

const ERROR_MESSAGE = 'Er kunnen door een technische storing helaas geen bouwdossiers worden getoond. Probeer het later nog eens.';

const ConstructionFiles = ({ setFileName, fileName, user, endpoint }) => {
  const [results, setResults] = React.useState(null);
  const [errorMessage, setErrorMessage] = React.useState(false);
  const [loading, setLoading] = React.useState(false);
  const [imageViewerActive, setImageViewerActive] = React.useState(false);

  const { trackPageView } = useMatomo();
  const { documentTitle, setDocumentTitle } = useDocumentTitle();

  const {
    titel: title,
    subdossiers,
    datering: date,
    dossier_type: fileType,
    dossiernr: fileNumber
  } = results || {};

  async function fetchConstructionFiles() {
    setLoading(true);
    try {
      const data = await getByUrl(endpoint);
      setResults(data);
    } catch (e) {
      setErrorMessage(ERROR_MESSAGE);
    }
    setLoading(false);
  }

  React.useEffect(() => {
    fetchConstructionFiles();
  }, []);

  // Effect to update the documentTitle
  React.useEffect(() => {
    if (title) {
      setDocumentTitle(imageViewerActive && 'Bouwtekening', [title]);
    }
  }, [title, imageViewerActive]);

  // Track pageView when documentTitle changes
  React.useEffect(() => {
    if (title) {
      trackPageView(documentTitle);
    }
  }, [documentTitle]);

  React.useEffect(() => {
    setImageViewerActive(!!fileName);
  }, [fileName]);

  const withGrid = (children) => (
    <GridContainer direction="column" gutterX={20} gutterY={20}>
      <GridItem>
        {children}
      </GridItem>
    </GridContainer>
  );
  const noResultsTemplate = withGrid(
    <Typography element="em">Geen resultaten gevonden</Typography>
  );

  const notAuthorizedTemplate = withGrid(
    <NotAuthorizedMessage />
  );

  const loadingTemplate = withGrid(
    <LoadingIndicator />
  );

  const resultsTemplate = (
    <div className="c-construction-files">
      {withGrid(
        <React.Fragment>
          <Typography
            className="c-construction-files__title"
            element="h3"
          >
            Bouwdossier
          </Typography>
          <Typography element="h1">{title}</Typography>
        </React.Fragment>
      )}

      <div className="c-ds-table">
        <div className="c-ds-table__body">
          <div className="c-ds-table__row">
            <div className="c-ds-table__cell">
              <div className="qa-table-value">Datering</div>
            </div>
            <div className="c-ds-table__cell">
              <div className="qa-table-value">{date}</div>
            </div>
          </div>
          <div className="c-ds-table__row">
            <div className="c-ds-table__cell">
              <div className="qa-table-value">Type</div>
            </div>
            <div className="c-ds-table__cell">
              <div className="qa-table-value">{fileType}</div>
            </div>
          </div>
          <div className="c-ds-table__row">
            <div className="c-ds-table__cell">
              <div className="qa-table-value">Dossier nummer</div>
            </div>
            <div className="c-ds-table__cell">
              <div className="qa-table-value">{fileNumber}</div>
            </div>
          </div>
        </div>
      </div>

      {subdossiers &&
      subdossiers.length &&
      subdossiers.map(({ bestanden, titel: subdossierTitle }) => (
        <Gallery
          key={subdossierTitle}
          title={subdossierTitle}
          thumbs={bestanden}
          onClick={setFileName}
          max={6}
        />
      ))}
    </div>
  );

  return user.scopes.includes(SCOPES['BD/R'])
    ? errorMessage ? <ErrorMessage errorMessage={errorMessage} /> : (
      <React.Fragment>
        {imageViewerActive &&
        <ImageViewer {...{ fileName, title }} contextMenu={<ContextMenu fileName={fileName} />} />}
        {loading && loadingTemplate}
        {(!loading && !fileName) && (results ? resultsTemplate : noResultsTemplate)}
      </React.Fragment>)
    : notAuthorizedTemplate;
};

ConstructionFiles.propTypes = {
  setFileName: PropTypes.func.isRequired,
  fileName: PropTypes.string,
  user: PropTypes.shape({}).isRequired,
  endpoint: PropTypes.string.isRequired
};

const mapStateToProps = (state) => ({
  fileName: getFileName(state),
  endpoint: `${SHARED_CONFIG.API_ROOT}stadsarchief/bouwdossier/${getLocationPayload(state).id.replace('id', '')}/`,
  user: getUser(state)
});

const mapDispatchToProps = (dispatch) => bindActionCreators({
  setFileName: setCurrentFile
}, dispatch);

export default connect(mapStateToProps, mapDispatchToProps)(ConstructionFiles);
