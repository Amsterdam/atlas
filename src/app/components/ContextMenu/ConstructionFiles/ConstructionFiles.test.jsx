import React from 'react'
import { shallow } from 'enzyme'
import configureMockStore from 'redux-mock-store'
import ConstructionFiles from './ConstructionFiles'
import { sharePage } from '../../../../shared/ducks/ui/ui'

jest.mock('../../../../shared/ducks/ui/ui')

const mockHandleDownload = jest.fn()

describe('ContextMenu for ConstructionFiles viewer', () => {
  let component
  const mockOpenPrintMode = jest.fn()

  beforeEach(() => {
    const props = {
      fileName: 'filename.jpg',
      openPrintMode: mockOpenPrintMode,
      isImage: true,
      handleDownload: mockHandleDownload,
    }
    const initialState = {
      map: {
        mapPanelActive: true,
      },
      ui: {
        viewMode: 'print',
      },
    }
    sharePage.mockImplementation(() => ({ type: 'action' }))

    const store = configureMockStore()({ ...initialState })
    component = shallow(<ConstructionFiles {...props} store={store} />).dive()
  })

  it('should render', () => {
    expect(component).toMatchSnapshot()
  })

  it('should handle the onClick events', () => {
    mockHandleDownload.mockClear()

    const downloadButton = component.find('ContextMenuItem')

    downloadButton.at(1).simulate('click')
    expect(mockHandleDownload).toHaveBeenCalledTimes(1)

    downloadButton.at(2).simulate('click')
    expect(mockHandleDownload).toHaveBeenCalledTimes(2)

    downloadButton.at(3).simulate('click')
    expect(mockHandleDownload).toHaveBeenCalledTimes(3)
  })
})
