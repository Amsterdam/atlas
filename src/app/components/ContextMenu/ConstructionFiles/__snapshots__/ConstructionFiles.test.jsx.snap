// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ContextMenu for ConstructionFiles viewer should render 1`] = `
<ContextMenu
  alt="Actiemenu"
  arrowIcon={<SvgChevronDown />}
  data-test="context-menu"
  icon={
    <Icon
      inline={true}
      padding={4}
      size={24}
    >
      <SvgEllipsis />
    </Icon>
  }
  position="bottom"
  tabindex={0}
>
  <ContextMenuItem
    icon={
      <Icon
        inline={true}
        padding={4}
        size={24}
      >
        <SvgPrint />
      </Icon>
    }
    onClick={[Function]}
    role="button"
  >
    Printen
  </ContextMenuItem>
  <ContextMenuItem
    as="button"
    disabled={false}
    download="filename.jpg_small"
    icon={
      <Icon
        inline={true}
        padding={4}
        size={24}
      >
        <SvgDownload />
      </Icon>
    }
    onClick={[Function]}
  >
    Download klein
  </ContextMenuItem>
  <ContextMenuItem
    as="button"
    disabled={false}
    download="filename.jpg_large"
    icon={
      <Icon
        inline={true}
        padding={4}
        size={24}
      >
        <SvgDownload />
      </Icon>
    }
    onClick={[Function]}
  >
    Download groot
  </ContextMenuItem>
  <ContextMenuItem
    as="button"
    disabled={false}
    divider={true}
    download="filename.jpg_original"
    icon={
      <Icon
        inline={true}
        padding={4}
        size={24}
      >
        <SvgDownload />
      </Icon>
    }
    onClick={[Function]}
  >
    Download origineel
  </ContextMenuItem>
  <ContextMenuItem
    icon={
      <Icon
        inline={true}
        padding={4}
        size={24}
      >
        <SvgFacebookPadded />
      </Icon>
    }
    key="1"
    onClick={[Function]}
    role="button"
  >
    Deel via 
    Facebook
  </ContextMenuItem>
  <ContextMenuItem
    icon={
      <Icon
        inline={true}
        padding={4}
        size={24}
      >
        <SvgTwitter />
      </Icon>
    }
    key="2"
    onClick={[Function]}
    role="button"
  >
    Deel via 
    Twitter
  </ContextMenuItem>
  <ContextMenuItem
    icon={
      <Icon
        inline={true}
        padding={4}
        size={24}
      >
        <SvgLinkedin />
      </Icon>
    }
    key="3"
    onClick={[Function]}
    role="button"
  >
    Deel via 
    Linkedin
  </ContextMenuItem>
  <ContextMenuItem
    icon={
      <Icon
        inline={true}
        padding={4}
        size={24}
      >
        <SvgEmail />
      </Icon>
    }
    key="4"
    onClick={[Function]}
    role="button"
  >
    Deel via 
    E-mail
  </ContextMenuItem>
</ContextMenu>
`;
